docs/git-style-guid.adoc

```diff
@@ -0,0 +1,288 @@
+----
++
+Ultimately, when writing a commit message, think about what you would
+need to know if you run across the commit in a year from now.
+* If a _commit A_ depends on _commit B_, the dependency should be stated
+in the message of _commit A_. Use the SHA1 when referring to commits.
++
+Similarly, if _commit A_ solves a bug introduced by _commit B_, it
+should also be stated in the message of _commit A_.
+* If a commit is going to be squashed to another commit use the
+`--squash` and `--fixup` flags respectively, in order to make the
+intention clear:
++
+[source,shell]
+----
+$ git commit --squash f387cab2
+----
++
+_(Tip: Use the `--autosquash` flag when rebasing. The marked commits
+will be squashed automatically.)_
+
+=== Merging
+
+* *Do not rewrite published history.* The repository’s history is
+valuable in its own right and it is very important to be able to tell
+_what actually happened_. Altering published history is a common source
+of problems for anyone working on the project.
+* However, there are cases where rewriting history is legitimate. These
+are when:
+** You are the only one working on the branch and it is not being
+reviewed.
+** You want to tidy up your branch (eg. squash commits) and/or rebase it
+onto the ``master'' in order to merge it later.
++
+That said, _never rewrite the history of the ``master'' branch_ or any
+other special branches (ie. used by production or CI servers).
+* Keep the history _clean_ and _simple_. _Just before you merge_ your
+branch:
+[arabic]
+. Make sure it conforms to the style guide and perform any needed
+actions if it doesn’t (squash/reorder commits, reword messages etc.)
+. Rebase it onto the branch it’s going to be merged to:
++
+[source,shell]
+----
+[my-branch] $ git fetch
+[my-branch] $ git rebase origin/master
+# then merge
+----
++
+This results in a branch that can be applied directly to the end of the
+``master'' branch and results in a very simple history.
++
+_(Note: This strategy is better suited for projects with short-running
+branches. Otherwise it might be better to occassionally merge the
+``master'' branch instead of rebasing onto it.)_
+* If your branch includes more than one commit, do not merge with a
+fast-forward:
++
+[source,shell]
+----
+# good - ensures that a merge commit is created
+$ git merge --no-ff my-branch
+
+# bad
+$ git merge my-branch
+----
+
+=== Misc.
+
+* There are various workflows and each one has its strengths and
+weaknesses. Whether a workflow fits your case, depends on the team, the
+project and your development procedures.
++
+That said, it is important to actually _choose_ a workflow and stick
+with it.
+* _Be consistent._ This is related to the workflow but also expands to
+things like commit messages, branch names and tags. Having a consistent
+style throughout the repository makes it easy to understand what is
+going on by looking at the log, a commit message etc.
+* _Test before you push._ Do not push half-done work.
```

Comment:
Good point.
What about: Don't mark your PR as "ready for review" before you tested it.  For that

- open it as draft
- or set `WIP` in the title

?

Reply:
I like drafts and WIP. Both are good I think.

Reply:
But I can push to my branch without opening a PR. I think that should still be encouraged.

Reply:
Yeah, I'm 100% for pushing commits often. I mean, not for every sentence I change. But I pushed half done work all the time. What if my laptop dies? What if I need someone to collaborate with me on something?

I don't want to confuse pushing commits to the branch you're working on with a PR though. I think we just need to fix the wording here to get the point across that you should not open a PR, even as WIP or draft, until it is pretty much done. Maybe you're still doing some testing or some final touches.

Reply:
I too like to push to my branch early and often, for collaboration.  Others can collaborate on a branch.  In the past I've also pushed into PRs.  Which has better collab tools, branches or PRs?

I think this would improve the signal to noise ratio in PRs.

Reply:
git-style-guide updated